### PR TITLE
859 explain empty intersection

### PR DESCRIPTION
**Pull request by ShiriMoran**

### CODE REVIEW

File: pkg/awsvpc/examples/out/explain_out/subnet_to_subnet_all_vpcs_explain_detail.txt [hunk 1]
```diff
@@ -22,21 +22,32 @@ Path is disabled; The relevant rules are:
...
 
 ------------------------------------------------------------------------------------------------------------------------
 
-Connections from r1[10.240.48.198] to q2[10.240.32.122]: protocol: TCP dst-ports: 9080
+No connections from r1[10.240.48.198] to q1[10.240.32.91];
+	connection is blocked since ingress's and egress's connections do not intersect.
+	egress connection: protocol: TCP dst-ports: 9080, ingress connection: protocol: UDP
+
+Egress: security group GroupId:22 allows connection; network ACL acl1 allows connection
+Ingress: network ACL acl1 allows connection; security group GroupId:15 allows connection
```
Comment (zivnevo) [inline pkg/awsvpc/examples/out/explain_out/subnet_to_subnet_all_vpcs_explain_detail.txt:44]:
```suggestion
No connectivity from r1[10.240.48.198] to q1[10.240.32.91];
	connectivity is blocked since traffic patterns allowed at ingress are disjoint from the traffic patterns allowed at egress.
	allowed egress traffic: protocol: TCP dst-ports: 9080, allowed ingress traffic: protocol: UDP

Egress: security group GroupId:22 allows connection; network ACL acl1 allows connection
Ingress: network ACL acl1 allows connection; security group GroupId:15 allows connection
```

File: pkg/vpcmodel/explainabilityConnectivity.go [hunk 2]
```diff
@@ -217,55 +220,55 @@ func (details *rulesAndConnDetails) computeRoutersAndFilters(c *VPCConfig) (err
...
 	for _, layer := range FilterLayers {
 		filterIsRelevant := filters[layer]
 		potentialRules := rulesLayers[layer]
-		// The filter is blocking if it is relevant and has no allow rules
-		// this computation is meaningful only when rulesLayers are allow rules and is ignored otherwise
-		if filterIsRelevant && !filterHasRelevantRules(potentialRules) {
-			directionEnabled = false
-		}
 		if filterIsRelevant {
+			conn = conn.Intersect(connOfLayer(potentialRules))
```
Comment (zivnevo) [inline pkg/vpcmodel/explainabilityConnectivity.go:258]:
```suggestion
		if filterIsRelevant {
			potentialRules := rulesLayers[layer]
			conn = conn.Intersect(connOfLayer(potentialRules))
```

File: pkg/vpcmodel/explainabilityPrint.go [hunk 3]
```diff
@@ -118,6 +120,8 @@ func explainMissingCrossVpcRouter(src, dst string, connQuery *connection.Set) st
 // the connection is blocked and only part of the path is printed then 2 is printed so that the relevant information
 // is provided regardless of where the is blocking
 // 4 is printed only in detailed mode
+//
+//nolint:gocyclo // better not split into two function
```
Comment (zivnevo) [inline pkg/vpcmodel/explainabilityPrint.go:124]: Consider breaking this function to sub-functions.
It got pretty huge and rather hard to follow.
e.g., put all code under `if verbose` in a different function

File: pkg/vpcmodel/explainabilityPrint.go [hunk 4]
```diff
@@ -127,8 +131,10 @@ func (g *groupedConnLine) explainabilityLineStr(c *VPCConfig, connQuery *connect
 	needEgress := !src.IsExternal()
 	needIngress := !dst.IsExternal()
 	loadBalancerBlocking := loadBalancerRule != nil && loadBalancerRule.Deny(false)
-	ingressBlocking := !expDetails.ingressEnabled && needIngress
-	egressBlocking := !expDetails.egressEnabled && needEgress
+	ingressEnabled := expDetails.ingressConn != nil && !expDetails.ingressConn.IsEmpty()
```
Comment (zivnevo) [inline pkg/vpcmodel/explainabilityPrint.go:134]: When can `ingressConn` be nil?

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/vpcmodel/explainabilityPrint.go (hunk 4 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityPrint.go:134]: When the connection is to the internet

File: pkg/vpcmodel/explainabilityConnectivity.go (hunk 2 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityConnectivity.go:258]: done

File: pkg/awsvpc/examples/out/explain_out/subnet_to_subnet_all_vpcs_explain_detail.txt (hunk 1 above)
Comment (ShiriMoran) [reply pkg/awsvpc/examples/out/explain_out/subnet_to_subnet_all_vpcs_explain_detail.txt]: done

**Review by zivnevo — APPROVED** [review]
